autoversioner/cmd/versiongenerator/main.go

```diff
@@ -0,0 +1,97 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"os"
+	"sort"
+
+	ghub "github.com/GoogleCloudPlatform/guest-test-infra/autoversioner/github"
+	"github.com/GoogleCloudPlatform/guest-test-infra/autoversioner/version"
+	"github.com/google/go-github/github"
+)
+
+var (
+	tokenFile = flag.String("token-file-path", "", "path to github token file")
+	org       = flag.String("org", "", "organization name")
+	repo      = flag.String("repo", "", "repository name")
+)
+
+func main() {
+	ctx := context.Background()
+	flag.Parse()
+	err := validateFlags()
+	if err != nil {
+		fmt.Printf("error validating flags: %+v", err)
+		os.Exit(1)
+	}
+
+	buildID, err := generateVersion(ctx)
+	if err != nil {
+		fmt.Printf("error generating build id: %+v", err)
+		os.Exit(1)
+	}
+
+	fmt.Printf("Generated version: %s\n", buildID)
+}
+
+// gets the latest buildID
+// fetches all the  tags associated with the org/repo
+// sorts the matching versiontags in order of increasing order
+// we have to do this because there is no way to get the latest
+// version
+func generateVersion(ctx context.Context) (string, error) {
+	fetcher, err := ghub.NewClient(ctx, *tokenFile)
+	if err != nil {
+		return "", fmt.Errorf("error creating fetcher: %+v", err)
+	}
+
+	tags, err := fetcher.ListTags(ctx, *org, *repo)
+	if err != nil {
+		return "", fmt.Errorf("error fetching tags: %+v", err)
+	}
+
+	versions, err := getVersions(tags)
+	// this repository is using our build and release pipeline
+	// for the first time
+	if len(versions) == 0 {
+		return version.FirstVersionToday().String(), nil
```

Comment:
since incrementver handles the day change, could we omit the duplicate logic in `firstversiontoday` and just set this like:
```
version := NonSemanticVer{}
if len(versions) != 0 {
  version = versions[len(versions)-1]
}
return version.IncrementVersion().String(), nil
```

or something similar? my overall feeling is we have two paths to creating the first version for the day, this one which is only invoked as first time for repo, and then the date-doesnt-match logic in incrementversion()

Reply:
got it. yeah, removed the redundancy.